database/002_initial_schema.sql

```diff
@@ -1,5 +1,470 @@
+ALTER TABLE companyuser ENABLE ROW LEVEL SECURITY;
+ALTER TABLE event ENABLE ROW LEVEL SECURITY;
+ALTER TABLE offer ENABLE ROW LEVEL SECURITY;
+ALTER TABLE wish ENABLE ROW LEVEL SECURITY;
+ALTER TABLE contracts ENABLE ROW LEVEL SECURITY;
+ALTER TABLE ad ENABLE ROW LEVEL SECURITY;
+ALTER TABLE queue ENABLE ROW LEVEL SECURITY;
+ALTER TABLE signatures ENABLE ROW LEVEL SECURITY;
+ALTER TABLE bookingevent ENABLE ROW LEVEL SECURITY;
+ALTER TABLE itex ENABLE ROW LEVEL SECURITY;
+
+-- Views 
+
+
+--Functions
+
+CREATE FUNCTION get_CID_companyUser(luser text) RETURNS INTEGER AS $$
+    SELECT CID
+    FROM companyuser
+    WHERE fullName = luser
+    $$ LANGUAGE SQL STRICT STABLE SECURITY DEFINER;
+CREATE FUNCTION get_CUID_companyUser(luser text) RETURNS INTEGER AS $$
+    SELECT CUID
+    FROM companyuser
+    WHERE fullName = luser
+    $$ LANGUAGE SQL STRICT STABLE SECURITY DEFINER;
+CREATE FUNCTION get_CID_company(luser text) RETURNS INTEGER AS $$
+    SELECT CID
+    FROM company
+    WHERE fullName = luser
+    $$ LANGUAGE SQL STRICT STABLE SECURITY DEFINER;
+
+-- Groups 
+
+CREATE GROUP companies WITH NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT;
+
+CREATE GROUP companyUsers WITH NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT;
+
+CREATE GROUP onlineUsers WITH NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT;
+
+
+-- companyUser_policy
+CREATE POLICY companyUser_select_policy
+    ON companyuser
+    FOR SELECT
+    USING (CID = get_CID_companyUser(current_user) or CID = get_CID_company(current_user))
+;
+CREATE POLICY companyUser_update_policy
+    ON companyuser
+    FOR UPDATE
+    USING (fullname = current_user)
+    WITH CHECK (fullname = current_user)
+;
+CREATE POLICY companyUser_insert_policy
+    ON companyuser
+    FOR INSERT
+    WITH CHECK( CID = get_CID_company(current_user))
+;
+CREATE POLICY companyUser_ow_policy
+    ON companyuser
+    FOR All
+    TO onlineUsers
+    USING (TRUE)
+    WITH CHECK(TRUE)
+;
+-- company_policy
+CREATE POLICY company_select_policy
+    ON company
+    FOR SELECT
+    USING (CID = get_CID_companyUser(current_user) or CID = get_CID_company(current_user))
+;
+CREATE POLICY company_company_policy
+    ON company
+    FOR All
+    TO companies
+    USING (CID = get_CID_company(current_user))
+    WITH CHECK(CID = get_CID_company(current_user))
+;
+CREATE POLICY company_ow_policy
+    ON company
+    FOR All
+    TO onlineUsers
+    USING (TRUE)
+    WITH CHECK(TRUE)
+;
+--Event
+CREATE POLICY event_ow_policy
+    ON event
+    FOR ALL
+    TO onlineUsers
+    USING (TRUE)
+    WITH CHECK (TRUE)
+;
+-- offer_policy
+CREATE POLICY offer_select_policy
+    ON offer
+    FOR SELECT
+    USING (CID = get_CID_companyUser(current_user) or CID = get_CID_company(current_user))
+;
+CREATE POLICY offer_update_policy
+    ON offer
+    FOR UPDATE
+    USING ((CID = get_CID_companyUser(current_user) or CID = get_CID_company(current_user))  and statusCode in ('1', '2'))
+    WITH CHECK (statusCode in ('1', '2'))
```

Comment:
I think they should be able to update an offer if they want to give us more money to be part of an event, or if they have reconsidered their offer i any other way.


The statusCode is meant to reflect the enum status at the top of the file. Being either a rejection, offer given etc. since we have not decided on the different statusCodes yet i have placed 1 and 2 as placeholders.

Reply:
We can leave it in for now